src/libstd/sys/unix/os.rs

```diff
@@ -328,6 +328,22 @@ pub fn current_exe() -> io::Result<PathBuf> {
 #[cfg(any(target_os = "linux", target_os = "android", target_os = "emscripten"))]
 pub fn current_exe() -> io::Result<PathBuf> {
     match crate::fs::read_link("/proc/self/exe") {
+        Ok(path) => {
+            let path_str = path.to_str().ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "path contains invalid UTF-8 data",
+            ))?;
+            unsafe {
+                if libc::access(path_str.as_ptr() as *const i8, libc::F_OK) == 0 {
```

Comment:
Does this FFI call require a NUL-terminated string? Rust strings are not.

Reply:
Either way `lstat` is the better way to test if the path indicated exists or not; access checks if the current user can also read the path. Though this gets weird if you do create a file with the ` (deleted)` suffix on in; this will detect it as existing, but the link is actually broken (`/proc` is, probably not surprisingly, magic here). I think `stat("/proc/self/exe")` may actually be a better solution for checking if the path exists.

I don't know; there's no good fool-proof way to do this; we're just trying to dodge the throwing knives someone is sending our way at this point.

Reply:
@clarfon outlined a possible approach, but that takes at least 4 syscalls. https://github.com/rust-lang/rust/issues/69343#issuecomment-590155375

I'm not sure if that's worth it because under race conditions paths must be stale the very moment you get them from a syscall. The only race-free way to do these things is to operate on file descriptors instead of paths.

Reply:
Yup, although I will say that the number of syscalls for this seems incredibly reasonable for an API that will be called extremely infrequently (likely seconds between invocations).

Reply:
Note that once the symlink is broken due to the target being deleted, it is always broken, so it's at least a one-way gate here. But that cannot rely on `readlink` followed by use of the path; it must be gotten from `stat("/proc/self/exe")` directly.